# 1
Utilize o conjunto de dados [Go To College Dataset](https://www.kaggle.com/datasets/saddamazyazy/go-to-college-dataset) ou um conjunto de dados de **classificação supervisionada** de sua escolha.

In [79]:
import pickle
import pandas as pd

In [80]:
with open("clf.pkl", "rb") as files:
    clf = pickle.load(files)

clf

KNeighborsClassifier(n_neighbors=10)

In [81]:
with open('X.pkl', 'rb') as files:
    X =pickle.load(files)

X 

type_school_encoded  school_accreditation_encoded  gender_encoded  \
0                      0                             0               1   
1                      0                             0               1   
2                      0                             1               0   
3                      1                             1               1   
4                      0                             0               0   
..                   ...                           ...             ...   
995                    1                             0               0   
996                    0                             1               0   
997                    1                             0               1   
998                    0                             1               1   
999                    0                             1               1   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  \
0                   1                  1                              0   
1                   1                  1                              0   
2                   4                  1                              0   
3                   4                  0                              1   
4                   4                  1                              0   
..                ...                ...                            ...   
995                 4                  0                              1   
996                 1                  0                              1   
997                 1                  1                              0   
998                 3                  0                              1   
999                 0                  1                              1   

     will_go_to_college_encoded  
0                             1  
1                             1  
2                             1  
3                             1  
4                             0  
..                          ...  
995                           1  
996                           1  
997                           1  
998                           1  
999                           0  

[1000 rows x 7 columns]

In [82]:
with open('y.pkl', 'rb') as files:
    y = pickle.load(files)

y 

[1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,


# 2
Utilize a técnica k-fold cross validation. Compare os resultados com o melhor resultado da Atividade 6.

In [83]:
from sklearn.model_selection import cross_validate
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import recall_score,precision_score, f1_score, accuracy_score, classification_report

In [84]:
scores = cross_validate(clf, X, y, scoring=['accuracy','f1','recall'],cv=10)
scores

{'fit_time': array([0.01011181, 0.00452209, 0.        , 0.00800037, 0.        ,
        0.00799298, 0.00792384, 0.00886607, 0.00807762, 0.0085578 ]),
 'score_time': array([0.01665854, 0.0167129 , 0.01607728, 0.00799561, 0.008008  ,
        0.        , 0.        , 0.        , 0.        , 0.00817585]),
 'test_accuracy': array([0.98, 0.98, 0.99, 0.99, 1.  , 1.  , 0.96, 0.99, 1.  , 0.98]),
 'test_f1': array([0.97959184, 0.97959184, 0.99009901, 0.98989899, 1.        ,
        1.        , 0.95833333, 0.98989899, 1.        , 0.98      ]),
 'test_recall': array([0.96, 0.96, 1.  , 0.98, 1.  , 1.  , 0.92, 0.98, 1.  , 0.98])}

In [85]:
print(f"Resultado usando k-fold: {scores['test_accuracy'].mean()} +- {scores['test_accuracy'].std()}")
print(f"Tempo de treinamento usando k-fold: {scores['fit_time'].mean()} +- {scores['fit_time'].std()}")
print(f"Tempo de teste usando k-fold: {scores['score_time'].mean()} +- {scores['score_time'].std()}")

Resultado usando k-fold: 0.9870000000000001 +- 0.011874342087037928
Tempo de treinamento usando k-fold: 0.006405258178710937 +- 0.003466939639875949
Tempo de teste usando k-fold: 0.007362818717956543 +- 0.006841772788344304


No modelo Hold out no exercicio 6 a acurácia foi 95%, já no K-fold foi de 98%, o que torna o K-fold mais preciso.





# 3
Realize experimentos utilizando normalização e padronização. Compare os resultados com o melhor resultado da Atividade 6.

In [86]:
X

type_school_encoded  school_accreditation_encoded  gender_encoded  \
0                      0                             0               1   
1                      0                             0               1   
2                      0                             1               0   
3                      1                             1               1   
4                      0                             0               0   
..                   ...                           ...             ...   
995                    1                             0               0   
996                    0                             1               0   
997                    1                             0               1   
998                    0                             1               1   
999                    0                             1               1   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  \
0                   1                  1                              0   
1                   1                  1                              0   
2                   4                  1                              0   
3                   4                  0                              1   
4                   4                  1                              0   
..                ...                ...                            ...   
995                 4                  0                              1   
996                 1                  0                              1   
997                 1                  1                              0   
998                 3                  0                              1   
999                 0                  1                              1   

     will_go_to_college_encoded  
0                             1  
1                             1  
2                             1  
3                             1  
4                             0  
..                          ...  
995                           1  
996                           1  
997                           1  
998                           1  
999                           0  

[1000 rows x 7 columns]

In [87]:
df = X[['type_school_encoded', 'school_accreditation_encoded', 'gender_encoded', 'interest_encoded',
       'residence_encoded', 'parent_was_in_college_encoded',
       'will_go_to_college_encoded']]

df

type_school_encoded  school_accreditation_encoded  gender_encoded  \
0                      0                             0               1   
1                      0                             0               1   
2                      0                             1               0   
3                      1                             1               1   
4                      0                             0               0   
..                   ...                           ...             ...   
995                    1                             0               0   
996                    0                             1               0   
997                    1                             0               1   
998                    0                             1               1   
999                    0                             1               1   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  \
0                   1                  1                              0   
1                   1                  1                              0   
2                   4                  1                              0   
3                   4                  0                              1   
4                   4                  1                              0   
..                ...                ...                            ...   
995                 4                  0                              1   
996                 1                  0                              1   
997                 1                  1                              0   
998                 3                  0                              1   
999                 0                  1                              1   

     will_go_to_college_encoded  
0                             1  
1                             1  
2                             1  
3                             1  
4                             0  
..                          ...  
995                           1  
996                           1  
997                           1  
998                           1  
999                           0  

[1000 rows x 7 columns]

In [88]:
X_train,X_test,y_train,y_test = train_test_split(X, y, train_size=0.9, random_state=1)
clf = KNeighborsClassifier(n_neighbors=7)
clf.fit(X_train,y_train)
y_pred = clf.predict(X_test)
accuracy_score(y_test,y_pred)

0.97

In [89]:
from sklearn.preprocessing import StandardScaler
scaleX = StandardScaler()
X_train = scaleX.fit_transform(X_train)
X_test = scaleX.fit_transform(X_test)

X_train

array([[-0.79772404, -1.05014463,  0.97151659, ..., -1.09566829,
         0.96290958,  0.98894994],
       [-0.79772404,  0.95224979,  0.97151659, ..., -1.09566829,
         0.96290958,  0.98894994],
       [-0.79772404,  0.95224979, -1.0293185 , ...,  0.91268499,
         0.96290958, -1.01117353],
       ...,
       [ 1.25356634, -1.05014463, -1.0293185 , ...,  0.91268499,
        -1.03851911,  0.98894994],
       [ 1.25356634,  0.95224979,  0.97151659, ...,  0.91268499,
        -1.03851911, -1.01117353],
       [ 1.25356634, -1.05014463, -1.0293185 , ..., -1.09566829,
         0.96290958, -1.01117353]])

# 4
Tente realizar alguma outra etapa pré-processamento/transformação dos dados. Compare os resultados com os resultados já obtidos.

In [90]:
from sklearn.pipeline import Pipeline

In [91]:
X_train,X_test,y_train,y_test = train_test_split(X, y, train_size=0.9, random_state=1)

In [92]:
X_train

type_school_encoded  school_accreditation_encoded  gender_encoded  \
335                    0                             0               1   
13                     0                             1               1   
911                    0                             1               0   
156                    1                             0               1   
428                    0                             0               1   
..                   ...                           ...             ...   
767                    1                             0               0   
72                     0                             1               0   
908                    1                             0               0   
235                    1                             1               1   
37                     1                             0               0   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  \
335                 3                  0                              1   
13                  4                  0                              1   
911                 1                  1                              1   
156                 3                  1                              0   
428                 3                  0                              1   
..                ...                ...                            ...   
767                 3                  1                              0   
72                  4                  0                              1   
908                 4                  1                              0   
235                 3                  1                              0   
37                  4                  0                              1   

     will_go_to_college_encoded  
335                           1  
13                            1  
911                           0  
156                           0  
428                           0  
..                          ...  
767                           1  
72                            0  
908                           1  
235                           0  
37                            0  

[900 rows x 7 columns]

In [93]:
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', KNeighborsClassifier(n_neighbors = 7))
    ])

In [94]:
X_train

type_school_encoded  school_accreditation_encoded  gender_encoded  \
335                    0                             0               1   
13                     0                             1               1   
911                    0                             1               0   
156                    1                             0               1   
428                    0                             0               1   
..                   ...                           ...             ...   
767                    1                             0               0   
72                     0                             1               0   
908                    1                             0               0   
235                    1                             1               1   
37                     1                             0               0   

     interest_encoded  residence_encoded  parent_was_in_college_encoded  \
335                 3                  0                              1   
13                  4                  0                              1   
911                 1                  1                              1   
156                 3                  1                              0   
428                 3                  0                              1   
..                ...                ...                            ...   
767                 3                  1                              0   
72                  4                  0                              1   
908                 4                  1                              0   
235                 3                  1                              0   
37                  4                  0                              1   

     will_go_to_college_encoded  
335                           1  
13                            1  
911                           0  
156                           0  
428                           0  
..                          ...  
767                           1  
72                            0  
908                           1  
235                           0  
37                            0  

[900 rows x 7 columns]

In [95]:
pipeline.fit(X_train, y_train)

Pipeline(steps=[('scaler', StandardScaler()),
                ('clf', KNeighborsClassifier(n_neighbors=7))])

In [96]:
pipeline.score(X_test, y_test)

0.99

In [97]:
y_pred = pipeline.predict(X_test)
accuracy_score(y_test,y_pred)

0.99

Usando o pipeline a acurácia deu um valor maior a acuracia foi de 97% para 99%